### PR TITLE
added spectral norm

### LINKED ISSUE #3351: Feature request: spectral normalization
### Discussed in https://github.com/google/flax/discussions/2902

<div type='discussions-op-text'>

<sup>Originally posted by **rdilip** February 23, 2023</sup>
Are spectral normalization layers something that anyone's working on at the moment? They're present in Haiku, so there's a template to work off of, and they'd be very useful (e.g., to enable things like https://www.tensorflow.org/tutorials/understanding/sngp). </div>

### PR DESCRIPTION
**Pull request by chiamp**
Resolves #3351.

Implemented `nn.SpectralNorm` as a layer wrapper. Check out the API documentation [here](https://flax--3335.org.readthedocs.build/en/3335/api_reference/flax.linen/layers.html#flax.linen.SpectralNorm).

Example usage:
```
class Foo(nn.Module):
  @nn.compact
  def __call__(self, x, train):
    x = nn.Dense(3)(x)
    # only spectral normalize the params of the second Dense layer
    x = nn.SpectralNorm(nn.Dense(4))(x, update_stats=train)
    x = nn.Dense(5)(x)
    return x

# init
x = jnp.ones((1, 2))
y = jnp.ones((1, 5))
model = Foo()
variables = model.init(jax.random.PRNGKey(0), x, train=False)
print(jax.tree_map(lambda x: x.shape, variables))
# {'batch_stats': {'SpectralNorm_0': {'Dense_1/kernel/sigma': (),
#                                     'Dense_1/kernel/u': (1, 4)}},
#  'params': {'Dense_0': {'bias': (3,), 'kernel': (2, 3)},
#             'Dense_1': {'bias': (4,), 'kernel': (3, 4)},
#             'Dense_2': {'bias': (5,), 'kernel': (4, 5)}}}

# train
def train_step(variables, x, y):
  def loss_fn(params):
    logits, updates = model.apply(
        {'params': params, 'batch_stats': variables['batch_stats']},
        x,
        train=True,
        mutable=['batch_stats'],
    )
    loss = jnp.mean(optax.l2_loss(predictions=logits, targets=y))
    return loss, updates

  (loss, updates), grads = jax.value_and_grad(loss_fn, has_aux=True)(
      variables['params']
  )
  return {
      'params': jax.tree_map(
          lambda p, g: p - 0.1 * g, variables['params'], grads
      ),
      'batch_stats': updates['batch_stats'],
  }, loss
for _ in range(10):
  variables, loss = train_step(variables, x, y)

# inference / eval
out = model.apply(variables, x, train=False)
```

### CODE REVIEW

File: flax/linen/normalization.py [hunk 1]
```diff
@@ -618,3 +641,208 @@ def __call__(self, x):
...
+        + '/'
+        + '/'.join((dict_key.key for dict_key in path[1:]))
+        + '/u'
+    )
+    u_var = self.variable(
+        'batch_stats',
+        u_var_name,
+        jax.random.normal,
+        self.make_rng('params')
+        if not self.has_variable('batch_stats', u_var_name)
```
Comment (levskaya) [inline flax/linen/normalization.py:812]: maybe we should allow the stats collection to be set by a constructor arg, and just defaulting to 'batch_stats'

**Review by levskaya — APPROVED** [review]
I just don't like the `_z_normalize_and_scale` rename here, otherwise good.  see other comment inline.

File: flax/linen/normalization.py [hunk 2]
```diff
@@ -394,7 +415,7 @@ def __call__(self, x):
         use_fast_variance=self.use_fast_variance,
     )
 
-    return _normalize(
+    return _z_normalize_and_scale(
```
[the remark this answers is not in the document]
Comment (chiamp) [reply flax/linen/normalization.py]: I reverted it back to `_normalize`

File: flax/linen/normalization.py (hunk 1 above)
Comment (chiamp) [reply flax/linen/normalization.py]: added
